client-js/app/elements/labrad-grapher.ts

```diff
@@ -134,41 +185,61 @@ export class LabradGrapher extends polymer.Base {
         trashed: false
       });
     }
+
     for (let dir of dirs) {
-      if (isIncluded(dir)) {
-        items.push({
-          id: 'dir:' + dir.name,
-          name: dir.name,
-          url: dir.url,
-          isDir: true,
-          isDataset: false,
-          isParent: false,
-          starred: isTagged(dir, 'star'),
-          trashed: isTagged(dir, 'trash')
-        });
-      }
+      this.newDir(dir);
     }
+
     for (let dataset of datasets) {
-      if (isIncluded(dataset)) {
-        items.push({
-          id: 'dataset:' + dataset.name,
-          name: dataset.name,
-          url: dataset.url,
-          isDir: false,
-          isDataset: true,
-          isParent: false,
-          starred: isTagged(dataset, 'star'),
-          trashed: isTagged(dataset, 'trash')
-        });
-      }
+      this.newDataset(dataset);
+    }
+  }
+
+  private addItemToList_(item: ListItem): void {
+    this.push('listItems_', item);
+    if (this.filterListItems_([item]).length == 1) {
+      this.push('listItems', item);
     }
-    return items;
+  }
+
+  private filterListItems_(items: Array<ListItem>) {
+    if (this.showTrash) {
+      return items.filter((x) => {
+        return (x.trashed || x.isParent);
+      });
+    }
+
+    if (this.showStars) {
+      return items.filter((x) => {
+        return (x.starred && !x.trashed) || x.isParent;
+      });
+    }
+
+    return items.filter((x) => {
+      return !x.trashed || x.isParent;
+    });
+  }
+
+  private applyListFilter_(): void {
+    // Will cause `iron-list` to jump to the top.
```

Comment:
You mean that the view scrolls to the top of the filtered list? Maybe clarify this comment a bit.